nni/compression/pytorch/compressor.py

```diff
@@ -38,12 +39,36 @@ def __init__(self, model, config_list, optimizer=None):
         optimizer: pytorch optimizer
             optimizer used to train the model
         """
+        self.optimizer = optimizer
+        if optimizer is not None:
+            self.optimizer_cls = optimizer.__class__
+            self.optimizer_param_groups = optimizer.state_dict()['param_groups']
+            # don't save 'params' in a 'param_group' in 'param_groups'
+            [group.pop('params') for group in self.optimizer_param_groups]
+        if model is not None and config_list is not None:
+            self.reconfig(model, config_list)
+
+    def reconfig(self, model, config_list):
+        """
+        Re-config the compressor.
+        """
         assert isinstance(model, torch.nn.Module)
         self.validate_config(model, config_list)
 
         self.bound_model = model
         self.config_list = config_list
-        self.optimizer = optimizer
+
+        if self.optimizer is not None:
+            if self.optimizer_cls.__name__ == 'SGD':
+                self.optimizer = self.optimizer_cls(model.parameters(), lr=0.001)
+            else:
+                self.optimizer = self.optimizer_cls(model.parameters())
+            groups = self.optimizer.param_groups
+            assert len(groups) == len(self.optimizer_param_groups)
+            for group, temp_group in zip(groups, self.optimizer_param_groups):
+                assert len(group) - 1 == len(temp_group)
+                for k, v in temp_group.items():
+                    group[k] = deepcopy(v)
```

Comment:
Because this step is using the original optimizer param groups to reset the optimizer new by us. We want the optimizer in each `reconfig` to be the same as the optimizer transfer by the user.

Reply:
But in line45 `self.optimizer_param_groups = optimizer.state_dict()['param_groups']`, `self.optimizer_param_groups` get the reference of optimizer param_groups. It means that the following `reconfig` process will give new optimizer the final changed param_groups in the first optimizer.

Reply:
yes, and actually we never use the incoming optimizer from the user, we just get its state_dict, and all optimizers we used are initialized in line62-65. In this case, if the user has some changes on the incoming optimizer before initializing the compressor, we will get the change in the state_dict, but I think it is intuitive. And another problem is we can't get the initial param from an optimizer instance.